src/main.ts

```diff
@@ -51,16 +52,18 @@ export function deactivate() {
 //   subscriptions.add(busySignalProvider)
 // }
 
-export async function consumeOutlineProvider(provider: OutlineProvider) {
-  subscriptions.add(/*  providerRegistryEntry */ outlineProviderRegistry.addProvider(provider))
+export function consumeOutlineProvider(provider: OutlineProvider): Disposable {
+  const prividerDisposable = outlineProviderRegistry.addProvider(provider)
+  subscriptions.add(/*  providerRegistryEntry */ prividerDisposable)
 
   // NOTE Generate (try) an outline after obtaining a provider for the current active editor
   // this initial outline is always rendered no matter if it is visible or not,
   // this is because we can't track if the outline tab becomes visible suddenly later,
   // or if the editor changes later once outline is visible
   // so we need to have an outline for the current editor
   // the following updates rely on the visibility
-  await getOutline()
+  getOutline()
```

Comment:
We should await this promise.

Reply:
I don't think asynchronous functions can be used here because we need to return `Disposable` instead of `Promise<Disposable>` ... maybe.
Here, I thought it was okay to just call the getoutline function without having to await.

Reply:
The client should be able to await the function. Is this because of backwards compatibility?

Reply:
I did because the [docs](https://flight-manual.atom.io/behind-atom/sections/interacting-with-other-packages-via-services/) say I need to return a Disposable.
But can we use async function actually? (I don't know that)

Reply:
There is no requirement for not using async functions.

Reply:
I just tried this locally. (I wish I had tried it sooner.😅)
I tried it locally and it didn't work with the async function. The reason is that if we use an asynchronous function, Promise\<Disposable\> will be the return value, so the return value will be ignored.

The relevant line is here. There is a conditional branch depending on whether the return value has a `dispose` method, but Promise <Disposable> does not have a `dispose` method.

https://github.com/atom/service-hub/blob/b450ce54176d7743644dfb5db0f1df7138cefa17/src/provider.coffee#L23-L25

I think it's enough to trigger the start of processing without adding await to the getOutline function.

In summary, we can't use async functions here. (In my understanding. If I make a mistake, please correct it.)

Reply:
The disposable should be disposed of by the client.

Reply:
I think disposables are handled by inside Atom and not passed on to clients.

Reply:
If we want to use async API, then it should be done by the client.

Reply:
The issue with not awaiting the show outline is a race condition. What if Atom disposes the disposable before the outline is shown? That would cause some errors.

Reply:
For clarity, client means a language package like atom-typescript, right?

The client cannot receive the disposable that is the return value of consumOutlineProvider. (The consumer receives the return value of the provider, but not the other way around.)
The disposble is destroyed when the client package is deactivated, and the provider is removed from the outlineProviderRegistry. We cannot defer package deactivation using asynchronous functions.
Requesting an outline to a deactivated package will result in an error. Instead of waiting for getOutline and then discarding asynchronously, we should immediately discard it synchronously to avoid getting an error.

Reply:
Hmm, right. I think should be fixed in Atom itself as soon as possible. Atom should allow us to await the things.

Reply:
I think it's okay to merge this as it is, but after reading your comment, I started to think that the following changes would work as well.
What do you think?

```ts
export function consumeOutlineProvider(provider: OutlineProvider): Disposable {
  const prividerDisposable = outlineProviderRegistry.addProvider(provider)
  subscriptions.add(prividerDisposable)
  return new Disposable(()=>{
    getOutline().finally(()=>{
      prividerDisposable.dispose()
    })
  })
}
```

Reply:
This would result in the client not working after the first getOutline call. See my last suggestion. We can use that for now